### ISSUE TITLE
RFC Caching

### ISSUE THREAD

**Issue by ahopkins** [sanic-org/sanic#2034]
## Background

Sanic has two opposing goals: speed and flexibility. While these goals often both strive for simplicity, they also are often competing in design. To handle a wide variety of scenarios, the server must run through a number of checks, parsing data from the request, etc.

I propose a new **opt-in** caching mechanism that can be configured (much like `strict_slashes` or `host`) at various levels, in order of precedent:

- route,
- blueprint, and
- application.

There are potentially different places that we can serve the cache from:

- after reading the request line (fastest time to response, but would ignore incoming headers and middleware)
- after reading headers (would ignore middleware)
- after retrieving the route (we now have access to the path, params, etc)

This last one potentially gives the developer the most opportunity to decide _how_ to cache a request. At its essence, the idea would be to capture the bytes being sent in the response to a key/value store, and serve those bytes back instead of executing the handlers. With some initial crude testing, this gain can be 15-20%.

## API

I'd suggest that the developer can control the caching in several ways:

```python

app = Sanic("CachedApp", cache=RouteCache.PATH)

app.config.ROUTE_CACHE = "headers"

bp = Blueprint("NoCaching", cache=None)

app.route("/super-cache", cache=RouteCache.RAW)(...)

bp.route("/super-cache", cache="headers)(...)
```

As you can see, I intentionally am mixing `Enum` with strings.

## Potential strategies

I think that the caching layer would be easy to add inside the new router, as a composable layer. By default we do nothing, therefore no performance change. When a route is cacheable, we inject the cache storage function into the handler, and also the appropriate location in the request/response cycle.

**Comment by sjsadowski** [comment]
I will weigh in here that in my opinion, flexibility should always be secondary to speed in the decision-making process. It shouldn't be discarded completely, but the trade-offs should be weighed appropriately,

With that being said, adding caching to increase responsiveness of the handlers is a good thing in my opinion and if we're making it completely optional so that the default behavior doesn't change (at least not until further down the road), this gets an upvote from me.

**Comment by ahopkins** [comment]
As a point of clarification, I agree. Speed is a primary concern that should generally trump others. However, Sanic _should not_ become an opinionated framework, and _should_ ultimately provide the developer the greatest amount of flexibility possible.

**Comment by sjsadowski** [comment]
Absolutely, but feature decisions will always carry a bias - I don't think that would make the framework opinionated, but would reflect our choice of implementing things. For this reason, I think that bias should lean toward speed, and I think caching is a good item for that.

**Comment by SkBlaz** [comment]
Hi!  First, great work with the library. Any idea what preferred caching solution currently is? (few years later?)
